### PR TITLE
Introduced FlowCopier extension point

### PR DESCRIPTION
**Pull request by jglick**
Needed for https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/10.

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public abstract void copy(FlowExecutionOwner original, FlowExecutionOwner copy) throws IOException, InterruptedException;
+
+    /**
+     * Convenience implementation that only operates on true builds.
+     */
+    public static abstract class ByRun extends FlowCopier {
+
+        /**
+         * Copies metadata between builds
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java]: :ant: Missing a period.

File: src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+import hudson.ExtensionPoint;
+import hudson.model.Queue;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import java.io.IOException;
+
+/**
+ * A way for plugins to copy metadata and associated files from one flow execution to another.
+ * Useful when a new execution is not being created from scratch, but is a kind of clone of another.
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java]: "Kind of clone"?

File: src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     * Copies any required metadata or files from one to another.
+     * @param original an initial build, typically complete
+     * @param copy a new build, typically not yet started
+     */
+    public abstract void copy(FlowExecutionOwner original, FlowExecutionOwner copy) throws IOException, InterruptedException;
+
+    /**
+     * Convenience implementation that only operates on true builds.
+     */
+    public static abstract class ByRun extends FlowCopier {
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java]: :ant:  ByRun is a rather unhelpful classname.   Perhaps CopyFlowByRun would be better (or similar).

**Comment by rsandell** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java (hunk 3 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java]: Well, `FlowCopier.ByRun` is how you normally refer to it.

Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java]: Still not a great name, at least for IDE autocomplete purposes.

File: src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java (hunk 2 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java]: Explained out-of-band.

**Comment by svanoort** [comment]
🐝  with noted nits.

**Comment by svanoort** [comment]
re-🐝

**Comment by abayer** [comment]
:bee:

**Pull request merged**